ibis/backends/tests/test_client.py

```diff
@@ -929,25 +929,7 @@ def test_self_join_memory_table(backend, con, monkeypatch):
             lambda: ds.dataset(pa.table({"a": ["a"], "b": [1]})),
             "df_arrow_dataset",
             marks=[
-                pytest.mark.notimpl(
-                    [
-                        "bigquery",
-                        "clickhouse",
-                        "duckdb",
-                        "exasol",
-                        "impala",
-                        "mssql",
-                        "mysql",
-                        "oracle",
-                        "polars",
-                        "postgres",
-                        "pyspark",
-                        "risingwave",
-                        "snowflake",
-                        "sqlite",
-                        "trino",
-                    ]
```

Comment:
I'm a bit concerned about this. Duckdb will push the queries down into the dataset, avoiding loading the full dataset in memory at any point. All other backends will go the `to_frame()`/`to_pyarrow()` route though, which will fully load the data into memory before passing it to create a temp table in the backend. For small data that's fine, but if a user has a Dataset instead of a Table, it's probably larger data. I'm a bit wary of automatically loading this into memory all at once.

A different option would be to error when registering `Dataset` backed `memtable` objects for backends that lack an efficient way to handle them, telling the user to instead call `ibis.memtable(dataset.to_table())` themselves (we could raise this in `to_frame()`/`to_pyarrow()` on the `DatasetProxy`). This makes the potential perf footgun known.

Reply:
Yeah, I think this makes sense.

Reply:
Should we add a `FootgunError` to `common.exceptions`?

Reply:
Nah, I'd just use one of the common ones for this. `UnsupportedOperationError` seems fine, as does just a simple `TypeError`.